module/zstd/zstd.c

```diff
@@ -0,0 +1,611 @@
+ * requested size, it will be released and a new cached entry will be allocated.
+ * If other pooled objects are detected without beeing used for 2 minutes, they
+ * will be released, too.
+ *
+ * The concept is that high frequency memory allocations of bigger objects are
+ * expensive. So if alot of work is going on, allocations will be kept for a
+ * while and can be reused in that timeframe.
+ *
+ * The scheduled release will be updated every time a object is reused.
+ */
+static void *
+zstd_mempool_alloc(struct zstd_pool *zstd_mempool, size_t size)
+{
+	int i;
+	struct zstd_pool *pool;
+	struct zstd_kmem *mem = NULL;
+
+	if (!zstd_mempool) {
+		return (NULL);
+	}
+
+	/* Seek for preallocated memory slot and free obsolete slots */
+	for (i = 0; i < ZSTD_POOL_MAX; i++) {
+		pool = &zstd_mempool[i];
+		if (mutex_tryenter(&pool->barrier)) {
+			/*
+			 * Check if objects fits the size, if so we take it and
+			 * update the timestamp.
+			 */
+			if (!mem && pool->mem && size <= pool->size) {
+				pool->timeout = gethrestime_sec() +
+				    ZSTD_POOL_TIMEOUT;
+				mem = pool->mem;
+				continue;
+			}
+
+			/* Free memory if object is older than 2 minutes */
+			if (pool->mem && gethrestime_sec() > pool->timeout) {
+				kmem_free(pool->mem, pool->size);
+				pool->mem = NULL;
+				pool->size = 0;
+				pool->timeout = 0;
+			}
+
+			mutex_exit(&pool->barrier);
+		}
+	}
+
+	if (mem) {
+		return (mem);
+	}
+
+	/* If no preallocated slot was found, try to fill in a new one */
+	for (i = 0; i < ZSTD_POOL_MAX; i++) {
+		pool = &zstd_mempool[i];
+		if (mutex_tryenter(&pool->barrier)) {
+			/* Object is free, try to allocate new one */
+			if (!pool->mem) {
+				mem = vmem_alloc(size, KM_SLEEP);
+				pool->mem = mem;
+
+				/* Allocation successfull? */
+				if (pool->mem) {
+					/* Keep track for later release */
+					mem->pool = pool;
+					pool->size = size;
+					mem->kmem_type = ZSTD_KMEM_POOL;
+					mem->kmem_size = size;
+				}
+			}
+
+			if (size <= pool->size) {
+				/* Update timestamp */
+				pool->timeout = gethrestime_sec() +
+				    ZSTD_POOL_TIMEOUT;
+
+				return (pool->mem);
+			}
+
+			mutex_exit(&pool->barrier);
+		}
+	}
+
+	/*
+	 * If the pool is full or the allocation failed, try lazy allocation
+	 * instead.
+	 */
+	if (!mem) {
+		mem = vmem_alloc(size, KM_NOSLEEP);
+		if (mem) {
+			mem->pool = NULL;
+			mem->kmem_type = ZSTD_KMEM_DEFAULT;
+			mem->kmem_size = size;
+		}
+	}
+
+	return (mem);
+}
+
+/* Mark object as released by releasing the barrier mutex */
+static void
+zstd_mempool_free(struct zstd_kmem *z)
+{
+	mutex_exit(&z->pool->barrier);
+}
+
+/* Convert internal stored zfs level enum to zstd level */
+static enum zio_zstd_levels
+zstd_cookie_to_enum(int32_t level)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(fastlevels); i++) {
+		if (fastlevels[i].cookie == level) {
+			return (fastlevels[i].level);
+		}
+	}
+
+	/* This shouldn't happen. Cause a panic. */
+	printk(KERN_ERR "%s:Invalid ZSTD level encountered: %d",
```

Comment:
IIRC `printk(KERN_ERR, ...` does not cause a panic: we should update the comment or crash the system properly.

Reply:
nice find! we don't want to panic here, since we fall back to the default level

Reply:
Maybe we should explain with a comment why it is safe to return a default value when an invalid compression level is provided: "_This shouldn't happen_" is usually followed by a `VERIFY()/ASSER()`, not `printk()`.

Reply:
Thats what @c0d3z3r0 meant to say...
The whole section is in error. It shouldn't be an KERN_ERR, it shouldn't it shouldn't cause a panic and the commenting is also wrong ;)